File: 025. Reverse Nodes in k-Group/solution1-n.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    if (head === null || head.next === null || k < 2) {
        return head;
    }
    
    var dummyHead = new ListNode(0);
    dummyHead.next = head;
    var runner = head;
    var counter = k;
    var prevTail = dummyHead;
    
    while (runner !== null) {
        counter--;
        var nextNode = runner.next;
        
        if (counter === 0) {
            counter = k;
            var reversedHead = reverse(prevTail.next, nextNode);
            var reversedTail = prevTail.next;
            
            reversedTail.next = nextNode;
            prevTail.next = reversedHead;
            
            prevTail = reversedTail;
        }
        
        runner = nextNode;
    }
    
    return dummyHead.next;
};

var reverse = function(head, tail) {
    var prevNode = null;
    
    while (head !== tail) {
        var nextNode = head.next;
        head.next = prevNode;
        
        prevNode = head;
        head = nextNode;
    }
    
    return prevNode;
};

File: Leet-234. Palindrome Linked List/solution1-n.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) {
            return true;
        }
        
        ListNode runner = head;
        ListNode walker = head;
        
        while (runner != null && runner.next != null) {
            runner = runner.next.next;
            walker = walker.next;
        }
        
        if (runner != null) {
            walker = walker.next;
        }
        
        ListNode reversedHead = reverse(walker);
        while (reversedHead != null) {
            if (head.val != reversedHead.val) {
                return false;
            }
            
            reversedHead = reversedHead.next;
            head = head.next;
        }
        
        return true;
    }
    
    public ListNode reverse(ListNode head) {
        ListNode prev = null;
        
        while (head != null) {
            ListNode next = head.next;
            head.next = prev;
            
            prev = head;
            head = next;
        }
        
        return prev;        
    }
} 